tripal_chado/src/Controller/ChadoOrganismFormElementController.php

```diff
@@ -0,0 +1,342 @@
+   *
+   * This allows the same query to be used for both autocomplete and select.
+   *
+   * @param string $string
+   *   The string to be autocompleted, used to limit the query.
+   *   The string "%" has the special meaning of return all records.
+   * @param array $options
+   *   The following keys are used:
+   *     match_operator - Either 'CONTAINS' (default) or 'STARTS_WITH'.
+   *     match_limit - Desired number of autocomplete matching names to suggest.
+   *
+   * @return ?\Drupal\pgsql\Driver\Database\pgsql\Select
+   *   A database query object
+   */
+  public static function getQuery(string $string, array $options): ?Select {
+
+    // Set defaults.
+    $options['match_operator'] ??= 'CONTAINS';
+    $options['match_limit'] ??= 10;
+
+    // Generate a query only if $string is at least a character
+    // long and result count is set to a value greater than 0.
+    $query = NULL;
+    if (strlen($string) > 0 && $options['match_limit'] > 0) {
+
+      $connection = \Drupal::service('tripal_chado.database');
+
+      // Transform string into a search pattern with wildcards.
+      // Only STARTS_WITH and CONTAINS are supported.
+      $condition_value = $string . '%';
+      if ($options['match_operator'] == 'CONTAINS') {
+        $condition_value = '%' . $condition_value;
+      }
+      $query = $connection->select('1:organism', 'BT');
+      $query->leftJoin('1:cvterm', 'T', '"BT".type_id = "T".cvterm_id');
+      $query->addField('BT', 'organism_id', 'pkey');
+      $query->addField('BT', 'abbreviation', 'abbreviation');
+      $query->addField('BT', 'common_name', 'common_name');
+      $query->addExpression("CONCAT_WS(' ', genus, species, name, infraspecific_name)", 'organism');
+
+      // A single "%" wildcard is used to indicate that we should return
+      // all records. This is used for a form select element.
+      if ($string != '%') {
+        $query->where("CONCAT_WS(' ', genus, species, name, infraspecific_name) ILIKE :value",
+            [':value' => $condition_value]);
+      }
+      $query->orderBy('organism', 'ASC');
+      if ($options['match_limit'] > 0) {
+        $query->range(0, $options['match_limit']);
+      }
+    }
+    return $query;
+  }
+
+  /**
+   * Fetch the pkey organism_id number, given an autocomplete value.
+   *
+   * The value includes a numeric ID in parentheses at the end of the string.
+   *
+   * @param string $value
+   *   A value from an autocomplete with the ID in parentheses at the end,
+   *   e.g. "Tripalus bogusii (ignored) (123)".
+   *
+   * @return int
+   *   Primary key ID number of the record, or 0 if an unparsable $value was
+   *   passed, which can happen if the user did not let the autocomplete
+   *   supply a value.
+   */
+  public static function getPkeyId(string $value): int {
+    $id = 0;
+
+    $matches = [];
+    if (preg_match('/\((\d+)\)/', $value, $matches)) {
+      $id = $matches[array_key_last($matches)];
+    }
```

Comment:
```suggestion
    $matches = [];
    if (preg_match('/^\d+$/', $value)) {
      $id = $value;
    }
    elseif (preg_match('/\((\d+)\)/', $value, $matches)) {
      $id = $matches[array_key_last($matches)];
    }
```
If we add this, then this same function can be called for both select and autocomplete, and we can simplify form processing